File: scripts/density_cluster.py
#!/usr/bin/env python3
"""Density clustering of spectral peaks into independent tracks.

Harmonic series collapse first (one sung/played note is one source), then
DBSCAN in (log-f0, harmonicity, centroid) with adaptive eps. A solo voice
with overtones spanning 80–2000 Hz is ONE track, not five Hz-band guesses.
"""
from __future__ import annotations

import json
import math
import sys
from pathlib import Path

def group_harmonic_funds(peaks: list[dict]) -> list[dict]:
    funds: list[dict] = []
    for p in sorted(peaks, key=lambda x: -x["db"]):
        best: dict | None = None
        best_cents = 35.0
        for g in funds:
            n = round(p["f"] / g["f0"])
            if n < 2 or n > 16:
                continue
            cents = abs(1200 * math.log2(p["f"] / (n * g["f0"])))
            if cents < best_cents:
                best = g
                best_cents = cents
        if best is not None:
            best["members"].append(p)
            best["db"] = max(best["db"], p["db"])
        else:
            funds.append({"f0": p["f"], "db": p["db"], "members": [p]})
    for g in funds:
        w = 0.0
        f_sum = 0.0
        for m in g["members"]:
            mag = 10 ** (m["db"] / 20)
            w += mag
            f_sum += m["f"] * mag
        g["centroid"] = f_sum / (w or 1.0)
        g["harm"] = min(1.0, (len(g["members"]) - 1) / 5.0)
        g["logF"] = math.log2(g["f0"])
        g["logC"] = math.log2(max(g["centroid"], 1.0))
    return funds


def feat(g: dict) -> tuple[float, float, float]:
    return (g["logF"] * 0.42, g["harm"] * 1.8, g["logC"] * 0.35)


def dist(a: tuple[float, ...], b: tuple[float, ...]) -> float:
    return math.sqrt(sum((x - y) ** 2 for x, y in zip(a, b)))


DBSCAN_MIN_PTS = 2
EPS_FLOOR = 0.28
EPS_CAP = 0.85
# eps is scaled *below* the median nearest-neighbor gap so two genuinely
# distinct fundamentals (already harmonic-folded) stay separate. The old
# 1.35× inflation made eps larger than the typical inter-source gap, which let
# the single-linkage expansion chain A~B~C into one cluster even when A and C
# were far apart in feature space.
EPS_NEIGHBOR_SCALE = 0.9


def density_cluster_funds(funds: list[dict]) -> list[dict]:
    if not funds:
        return []
    pts = [{"g": g, "x": feat(g)} for g in funds]
    gaps: list[float] = []
    for i, p in enumerate(pts):
        best = 1e9
        for j, q in enumerate(pts):
            if i == j:
                continue
            d = dist(p["x"], q["x"])
            if d < best:
                best = d
        if best < 1e9:
            gaps.append(best)
    gaps.sort()
    median = gaps[len(gaps) // 2] if gaps else EPS_FLOOR
    eps = max(EPS_FLOOR, min(EPS_CAP, median * EPS_NEIGHBOR_SCALE))
    n = len(pts)
    labels = [-1] * n

    def neighbors(i: int) -> list[int]:
        return [j for j in range(n) if dist(pts[i]["x"], pts[j]["x"]) <= eps]

    # Proper DBSCAN: only core points (>= minPts points incl. self within eps)
    # seed a cluster. The previous flood fill treated every point as a core
    # point, i.e. single-linkage chaining.
    core = [len(neighbors(i)) + 1 >= DBSCAN_MIN_PTS for i in range(n)]
    cid = 0
    for i in range(n):
        if labels[i] != -1 or not core[i]:
            continue
        labels[i] = cid
        seed = list(neighbors(i))
        s = 0
        while s < len(seed):
            j = seed[s]
            s += 1
            if labels[j] == -1:
                labels[j] = cid
                if core[j]:
                    for k in neighbors(j):
                        if k not in seed:
                            seed.append(k)
        cid += 1

    # Isolated funds (no core neighbor) still become their own track.
    for i in range(n):
        if labels[i] == -1:
            labels[i] = cid
            cid += 1

    buckets: dict[int, list] = {}
    for i, lab in enumerate(labels):
        buckets.setdefault(lab, []).append(pts[i]["g"])
    clusters = []
    for members in buckets.values():
        members.sort(key=lambda g: -g["db"])
        head = members[0]
        clusters.append(
            {
                "f0": head["f0"],
                "db": max(g["db"] for g in members),
                "harm": sum(g["harm"] for g in members) / len(members),
                "n": sum(len(g["members"]) for g in members),
            }
        )
    clusters.sort(key=lambda c: -c["db"])
    return [c for c in clusters if c["db"] > -90]


def funds_as_clusters(funds: list[dict]) -> list[dict]:
    clusters = [
        {
            "f0": g["f0"],
            "db": g["db"],
            "harm": g["harm"],
            "n": len(g["members"]),
        }
        for g in funds
        if g["db"] > -90
    ]
    clusters.sort(key=lambda c: -c["db"])
    return clusters


def cluster_peaks(peaks: list[dict], *, merge_nearby: bool = True) -> list[dict]:
    funds = group_harmonic_funds(peaks)
    if not merge_nearby:
        return funds_as_clusters(funds)
    return density_cluster_funds(funds)


def heuristic_label(f0: float, harm: float) -> str:
    """Same nouns as iOS/web ClusterLabeler — voix is first-class."""
    if harm < 0.18 and f0 > 180:
        return "bruit"
    if f0 < 90:
        return "grave"
    if f0 < 280 and harm >= 0.35:
        return "voix"
    if f0 < 450:
        return "corps"
    if harm >= 0.55:
        return "nylon"
    if f0 > 1400:
        return "air"
    return ""


def tone_stack(f0: float, db0: float = -20.0, n_harm: int = 6) -> list[dict]:
    return [{"f": f0 * n, "db": db0 - 4.0 * (n - 1)} for n in range(1, n_harm + 1)]


def fixture_cases() -> list[dict]:
    return [
        {
            "name": "solo220",
            "peaks": tone_stack(220.0, -16.0, 8),
            "merge_nearby": True,
            "expect_n": 1,
            "expect_f0": [220.0],
        },
        {
            "name": "two_sources",
            "peaks": tone_stack(110.0, -18.0, 5) + tone_stack(523.25, -22.0, 4),
            "merge_nearby": True,
            "expect_n": 2,
            "expect_f0": [110.0, 523.25],
        },
        {
            "name": "ceg_offline",
            "peaks": tone_stack(130.81, -18.0, 4)
            + tone_stack(164.81, -20.0, 4)
            + tone_stack(196.0, -21.0, 3),
            "merge_nearby": False,
            "expect_n": 3,
            "expect_f0": [130.81, 164.81, 196.0],
        },
    ]


def write_fixtures(path: Path) -> None:
    cases = []
    for raw in fixture_cases():
        got = cluster_peaks(raw["peaks"], merge_nearby=raw["merge_nearby"])
        if len(got) != raw["expect_n"]:
            raise SystemExit(f"fixture {raw['name']} expected {raw['expect_n']} got {len(got)}")
        cases.append(raw)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps({"cases": cases}, indent=2), encoding="utf-8")


def main() -> None:
    solo = cluster_peaks(tone_stack(220.0, -16.0, 8))
    if len(solo) != 1:
        raise SystemExit(f"solo 220 Hz + harmonics must be 1 track, got {len(solo)}")

    voice_span = cluster_peaks(tone_stack(82.4, -16.0, 12))
    if len(voice_span) != 1:
        raise SystemExit(
            f"one harmonic series 82–990 Hz must be 1 track (not five bands), got {len(voice_span)}"
        )

    two = cluster_peaks(tone_stack(110.0, -18.0, 5) + tone_stack(523.25, -22.0, 4))
    if len(two) != 2:
        raise SystemExit(
            f"two independent sources must be 2 tracks, got {len(two)} {[round(c['f0'], 1) for c in two]}"
        )

    # Two distinct sources where one has an exact octave harmonic: 200 Hz must
    # fold into 100 Hz, leaving 141 Hz (inharmonic, ~600 cents up) as its own
    # track. Single-linkage used to chain these into one cluster.
    distinct = cluster_peaks(
        [{"f": 100.0, "db": -20.0}, {"f": 200.0, "db": -24.0}, {"f": 141.0, "db": -22.0}]
    )
    if len(distinct) != 2:
        raise SystemExit(
            f"octave-folded source + inharmonic neighbor must be 2 tracks, "
            f"got {len(distinct)} {[round(c['f0'], 1) for c in distinct]}"
        )

    chordish = cluster_peaks(
        tone_stack(130.81, -18.0, 4) + tone_stack(164.81, -20.0, 4) + tone_stack(196.0, -21.0, 3)
    )
    if not 1 <= len(chordish) <= 3:
        raise SystemExit(f"one-source chord should stay small, got {len(chordish)}")
    chord_off = cluster_peaks(
        tone_stack(130.81, -18.0, 4) + tone_stack(164.81, -20.0, 4) + tone_stack(196.0, -21.0, 3),
        merge_nearby=False,
    )
    if len(chord_off) != 3:
        raise SystemExit(
            f"offline analysis must keep C–E–G as 3 funds, got {len(chord_off)} "
            f"{[round(c['f0'], 1) for c in chord_off]}"
        )
    if len(chordish) >= 5:
        raise SystemExit("must not invent five instrument tracks for one chord")

    # Psytrance-like: kick series + inharmonic highs (n>8 so they are not harmonics).
    psy = cluster_peaks(
        tone_stack(55.0, -12.0, 3)
        + [
            {"f": 2100.0, "db": -18.0},
            {"f": 2450.0, "db": -20.0},
            {"f": 3120.0, "db": -22.0},
            {"f": 3800.0, "db": -24.0},
        ]
    )
    if len(psy) < 2:
        raise SystemExit(f"kick + high synth must stay ≥2 sources, got {len(psy)}")
    if not any(c["f0"] < 80 for c in psy) or not any(c["f0"] > 1800 for c in psy):
        raise SystemExit(f"psytrance must keep a low and a high source: {[round(c['f0']) for c in psy]}")

    print(
        f"solo={len(solo)} voice_span={len(voice_span)} two={len(two)} "
        f"chord={len(chordish)} psy={len(psy)} distinct={len(distinct)}"
    )
    fixtures = Path(__file__).resolve().parents[1] / "piano" / "cluster_fixtures.json"
    write_fixtures(fixtures)
    print(f"density_cluster: OK wrote {fixtures}")


if __name__ == "__main__":
    try:
        main()
    except SystemExit as e:
        print(e, file=sys.stderr)
        raise

File: ios/CrayonPiano.swiftpm/DensityCluster.swift
import Foundation

struct SpecPeak {
    var f: Double
    var db: Float
}

struct SpectralCluster {
    var f0: Double
    var db: Float
    var harm: Double
    var centroid: Double
}

enum DensityCluster {
    /// Ports scripts/density_cluster.py. Shared cases live in piano/cluster_fixtures.json.
    /// No count cap — every independent fund stays a source. Harmonics already folded.

    private static let dbscanMinPts = 3
    private static let epsFloor = 0.28
    private static let epsCap = 0.85
    private static let epsNeighborScale = 0.9
    /// Distinct fundamentals more than this many cents apart never share a cluster.
    private static let minF0Cents = 70.0

    static func cluster(peaks: [SpecPeak], mergeNearby: Bool = true) -> [SpectralCluster] {
        let funds = groupHarmonicFunds(peaks)
        if !mergeNearby { return fundsAsClusters(funds) }
        return densityClusterFunds(funds)
    }

    private static func fundsAsClusters(_ funds: [Fund]) -> [SpectralCluster] {
        funds
            .filter { $0.db > -90 }
            .sorted { $0.db > $1.db }
            .map { SpectralCluster(f0: $0.f0, db: $0.db, harm: $0.harm, centroid: $0.centroid) }
    }

    private struct Fund {
        var f0: Double
        var db: Float
        var members: [SpecPeak]
        var centroid: Double = 0
        var harm: Double = 0
        var logF: Double = 0
        var logC: Double = 0
    }

    private static func cents(_ a: Double, _ b: Double) -> Double {
        if a <= 0 || b <= 0 { return 1e9 }
        return abs(1200 * log2(a / b))
    }

    private static func fillFundStats(_ g: inout Fund) {
        var w = 0.0
        var fSum = 0.0
        for m in g.members {
            let mag = pow(10.0, Double(m.db) / 20.0)
            w += mag
            fSum += m.f * mag
        }
        g.centroid = fSum / (w == 0 ? 1 : w)
        g.harm = min(1, Double(g.members.count - 1) / 5)
        g.logF = log2(max(g.f0, 1e-6))
        g.logC = log2(max(g.centroid, 1))
    }

    /// Prefer the lowest candidate that explains the partials (not the loudest).
    /// Half-f0 only when a peak sits there or an odd partial needs that f0.
    private static func refineFundF0(_ g: inout Fund) {
        let freqs = g.members.map(\.f).filter { $0 > 0 }
        guard !freqs.isEmpty else { return }
        var cands = freqs
        cands.append(contentsOf: freqs.map { $0 / 2 })
        var best = g.f0
        var bestScore = -1e18
        for cand in cands {
            if cand < 20 { continue }
            var hits = 0
            var oddHi = 0
            var hasSelf = false
            for f in freqs {
                let n = (f / cand).rounded()
                if n >= 1 && n <= 16 && cents(f, n * cand) < 35 {
                    hits += 1
                    if n == 1 { hasSelf = true }
                    if n >= 3 && n.truncatingRemainder(dividingBy: 2) == 1 { oddHi += 1 }
                }
            }
            if hits < 2 { continue }
            if !hasSelf && oddHi < 1 { continue }
            let score = Double(hits) * 1000 + Double(oddHi) * 10 - cand
            if score > bestScore {
                bestScore = score
                best = cand
            }
        }
        g.f0 = best
    }

    private static func mergeOctaveFunds(_ funds: [Fund]) -> [Fund] {
        let sorted = funds.sorted { $0.f0 < $1.f0 }
        var used = Set<Int>()
        var out: [Fund] = []
        for i in sorted.indices {
            if used.contains(i) { continue }
            var a = sorted[i]
            if i + 1 < sorted.count {
                for j in (i + 1)..<sorted.count {
                    if used.contains(j) { continue }
                    let b = sorted[j]
                    let n = a.f0 == 0 ? 0 : (b.f0 / a.f0).rounded()
                    if n < 1 || n > 8 { continue }
                    if cents(b.f0, n * a.f0) < 35 {
                        a.members.append(contentsOf: b.members)
                        a.db = max(a.db, b.db)
                        used.insert(j)
                    }
                }
            }
            fillFundStats(&a)
            out.append(a)
        }
        return out
    }

    private static func groupHarmonicFunds(_ peaks: [SpecPeak]) -> [Fund] {
        var funds: [Fund] = []
        for p in peaks.sorted(by: { $0.db > $1.db }) {
            var bestIdx: Int?
            var bestCents = 35.0
            for i in funds.indices {
                let n = (p.f / funds[i].f0).rounded()
                if n < 2 || n > 16 { continue }
                let c = cents(p.f, n * funds[i].f0)
                if c < bestCents {
                    bestIdx = i
                    bestCents = c
                }
            }
            if let idx = bestIdx {
                funds[idx].members.append(p)
                funds[idx].db = max(funds[idx].db, p.db)
            } else {
                funds.append(Fund(f0: p.f, db: p.db, members: [p]))
            }
        }
        for i in funds.indices {
            refineFundF0(&funds[i])
            fillFundStats(&funds[i])
        }
        return mergeOctaveFunds(funds)
    }

    private static func feat(_ g: Fund) -> (Double, Double, Double) {
        (g.logF * 0.42, g.harm * 1.8, g.logC * 0.35)
    }

    private static func dist(_ a: (Double, Double, Double), _ b: (Double, Double, Double)) -> Double {
        let dx = a.0 - b.0
        let dy = a.1 - b.1
        let dz = a.2 - b.2
        return sqrt(dx * dx + dy * dy + dz * dz)
    }

    private static func densityClusterFunds(_ funds: [Fund]) -> [SpectralCluster] {
        guard !funds.isEmpty else { return [] }
        let pts = funds.map { (g: $0, x: feat($0)) }
        var gaps: [Double] = []
        for i in pts.indices {
            var best = 1e9
            for j in pts.indices where i != j {
                let d = dist(pts[i].x, pts[j].x)
                if d < best { best = d }
            }
            if best < 1e9 { gaps.append(best) }
        }
        gaps.sort()
        let median = gaps.isEmpty ? epsFloor : gaps[gaps.count / 2]
        let eps = max(epsFloor, min(epsCap, median * epsNeighborScale))
        let n = pts.count
        var labels = Array(repeating: -1, count: n)

        func neighbors(_ i: Int) -> [Int] {
            var out: [Int] = []
            for j in 0..<n {
                if i == j { continue }
                if abs(1200 * (pts[i].g.logF - pts[j].g.logF)) > minF0Cents { continue }
                if dist(pts[i].x, pts[j].x) <= eps { out.append(j) }
            }
            return out
        }

        // neighbors exclude self; +1 counts the point. minPts=3 means a pair of
        // notes is not a core (avoids single-linkage A~B~C chains).
        var core = Array(repeating: false, count: n)
        for i in 0..<n { core[i] = neighbors(i).count + 1 >= dbscanMinPts }

        var cid = 0
        for i in 0..<n {
            if labels[i] != -1 || !core[i] { continue }
            labels[i] = cid
            var seed = neighbors(i)
            var s = 0
            while s < seed.count {
                let j = seed[s]
                s += 1
                if labels[j] == -1 {
                    labels[j] = cid
                    if core[j] {
                        for k in neighbors(j) where !seed.contains(k) {
                            seed.append(k)
                        }
                    }
                }
            }
            cid += 1
        }
        for i in 0..<n where labels[i] == -1 {
            labels[i] = cid
            cid += 1
        }

        var buckets: [Int: [Fund]] = [:]
        for i in 0..<n {
            buckets[labels[i], default: []].append(pts[i].g)
        }
        var clusters: [SpectralCluster] = []
        for members in buckets.values {
            let sorted = members.sorted { $0.db > $1.db }
            guard let head = sorted.first else { continue }
            let db = sorted.map(\.db).max() ?? head.db
            let harm = sorted.map(\.harm).reduce(0, +) / Double(sorted.count)
            clusters.append(SpectralCluster(f0: head.f0, db: db, harm: harm, centroid: head.centroid))
        }
        clusters.sort { $0.db > $1.db }
        return clusters.filter { $0.db > -90 }
    }
}
